### PR TITLE
Add rate limit to backend-listener

### PR DESCRIPTION
**Pull request by jjaferson**
# Description

We needed to apply the rate-limiting settings to the backend-listener to enable the usage of self-managed APICast, this PR also refactors the 3scale reconciler by moving many parts of rate limit resources creation to other package and go files.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Verification steps

1. Install RHOAM from the branch 
2. When the install is complete make sure the backend-listener pod has also an envoy sidecar container
3. Verify if an EnvoyConfig CR was created for the backend-listener
4. Ensure the backend-listener-proxy service was created in the 3scale namespace and the backend-listener route points to this service
5. Stop the operator and change the ratelimit configmap in the marin3r namespace to allow 2 requests
6. Curl the backend-listener external route one time, then curl the APICast route two times, when the APICast route responds to the second request, it should return `429 - too many requests`

### CODE REVIEW

**Comment by makslion** [comment]
Verified following: 
backend-listener pod has an envoy sidecar container
Screenshot from 2021-05-20 11-23-04
EnvoyConfig CR was created for the backend-listener
Screenshot from 2021-05-20 11-23-13
backend-listener-proxy service was created in the 3scale namespace and the backend-listener route points to this service
Screenshot from 2021-05-20 11-34-22

File: pkg/products/threescale/envoyconfig.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+	   - config:
+	       domain: apicast-ratelimit
+	       rate_limit_service:
+	         grpc_service:
+	           envoy_grpc:
+	             cluster_name: ratelimit
+	           timeout: 2s
+	       stage: 0
+	     name: envoy.rate_limit
+*/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:66]: shouldn't this commented out block be removed? :thinking:

File: pkg/products/threescale/envoyconfig.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+			},
+		},
+	},
+}
+
+/**
+ httpFilters:
+	- &tsHTTPRateLimitFilter
+	- name: envoy.router
+**/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:125]: as well as these lines. :thinking:

File: pkg/products/threescale/envoyconfig.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+		domain: apicast-ratelimit
+		rate_limit_service:
+		grpc_service:
+			envoy_grpc:
+			cluster_name: ratelimit
+			timeout: 2s
+		stage: 0
+	  name: envoy.filters.http.ratelimit
+	- name: envoy.filters.http.router
+**/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:158]: this is reflected in the code as well. Do we need it to be here? :thinking:

File: pkg/products/threescale/envoyconfig.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+func getBackendListenerHTTPFilters() ([]*hcm.HttpFilter, error) {
+
+	// function envoy_on_response(response_handle)
+	// 	rate_limit = response_handle:headers():get("x-envoy-ratelimited")
+	// 	if rate_limit ~= nil then
+	// 		response_handle:headers():add("3scale-rejection-reason", "limits_exceeded")
+	// 	end
+	// end
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:166]: duplicate of comment block above

File: pkg/products/threescale/envoyconfig.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	- match:
+		prefix: /
+		route:
+		cluster: apicast-ratelimit
+		rateLimits:
+		- actions:
+			- genericKey:
+				descriptorValue: slowpath
+			stage: 0
+*/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:207]: Should it be here? :thinking:

File: pkg/products/threescale/envoyconfig.go [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+virtual_hosts:
+	- name: backend-listener-ratelimit
+	domains: ["*"]
+	routes:
+		- match:
+			prefix: "/"
+		route:
+			cluster: backend-listener-ratelimit
+			rate_limits:
+**/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:244]: reflected in the code above

File: pkg/products/threescale/envoyconfig.go [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        - name: envoy.http_connection_manager
+          typedConfig:
+            '@type': type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
+			httpFilters:
+				&httpFilters
+            routeConfig:
+               name: local_route
+			   virtualHosts: virtualHosts
+			statPrefix: ingress_http
+**/
```
Comment (makslion) [inline pkg/products/threescale/envoyconfig.go:284]: do we need this? :thinking:

File: pkg/products/threescale/reconciler.go [hunk 8]
```diff
@@ -2411,3 +2022,203 @@ func (r *Reconciler) ensureDeploymentConfigsReady(ctx context.Context, serverCli
...
+
+func (r *Reconciler) getRateLimitServiceCR(ctx context.Context, serverClient k8sclient.Client) (*corev1.Service, error) {
+	rateLimitService := &corev1.Service{}
+	marin3rConfig, err := r.ConfigManager.ReadMarin3r()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load marin3r config in 3scale reconciler: %v", err)
+	}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{
+		Namespace: marin3rConfig.GetNamespace(),
+		Name:      "ratelimit",
```
Comment (makslion) [inline pkg/products/threescale/reconciler.go:2158]: only a single usage, but still a hardcoded value. Not sure should be changed to a constant/var tho :thinking:

File: pkg/products/threescale/reconciler.go [hunk 9]
```diff
@@ -2411,3 +2022,203 @@ func (r *Reconciler) ensureDeploymentConfigsReady(ctx context.Context, serverCli
...
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, serverClient, backendListenerService, func() error {
+		owner.AddIntegreatlyOwnerAnnotations(backendListenerService, r.installation)
+		backendListenerService.Spec.Ports = []corev1.ServicePort{
+			{
+				Name:       "http",
+				Protocol:   "TCP",
+				Port:       3001,
+				TargetPort: intstr.FromInt(3001),
```
Comment (makslion) [inline pkg/products/threescale/reconciler.go:2193]: those might as well be defined up top as constant

Comment (makslion) [inline pkg/products/threescale/reconciler.go:2193]: those can also be defined as constant up top, probably :thinking:

File: pkg/resources/ratelimit/envoy.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+		)
+		return phase, nil
+	}
+
+	if dc.Spec.Template.Labels == nil {
+		dc.Spec.Template.SetLabels(make(map[string]string))
+	}
+	if dc.Spec.Template.Annotations == nil {
+		dc.Spec.Template.SetAnnotations(make(map[string]string))
+	}
```
Comment (makslion) [inline pkg/resources/ratelimit/envoy.go:86]: could dc.Spec and dc.Spec.Template also be nils at this point?

File: pkg/resources/ratelimit/envoyconfig.go [hunk 11]
```diff
@@ -50,3 +71,187 @@ func DeleteEnvoyConfigsInNamespace(ctx context.Context, client k8sclient.Client,
...
+}
+
+/**
+  envoyAPI: v2
+  envoyResources:
+	clusters:
+	listeners:
+	nodeID:
+	serialization: yaml
+**/
```
Comment (makslion) [inline pkg/resources/ratelimit/envoyconfig.go:96]: should we have it here? :thinking:

File: pkg/resources/ratelimit/envoyconfig.go [hunk 12]
```diff
@@ -50,3 +71,187 @@ func DeleteEnvoyConfigsInNamespace(ctx context.Context, client k8sclient.Client,
...
+		endpoints:
+		- lbEndpoints:
+			- endpoint:
+				address:
+				socketAddress:
+					address: containerAddress
+					portValue: containerPort
+		name: apicast-ratelimit
+		type: STRICT_DNS
+**/
```
Comment (makslion) [inline pkg/resources/ratelimit/envoyconfig.go:176]: this is reflected in the code below. Shall it be here? :thinking:

File: pkg/resources/ratelimit/envoyconfig.go [hunk 13]
```diff
@@ -50,3 +71,187 @@ func DeleteEnvoyConfigsInNamespace(ctx context.Context, client k8sclient.Client,
...
+  - name: listenerName
+    value: |
+      address:
+        socketAddress:
+          address: listenerAddress
+          portValue: listenerPort
+      filterChains:
+      - filters: &filters
+	name: http
+*/
```
Comment (makslion) [inline pkg/resources/ratelimit/envoyconfig.go:224]: should it be here? :thinking:

File: pkg/resources/ratelimit/envoyconfig.go [hunk 14]
```diff
@@ -50,3 +71,187 @@ func DeleteEnvoyConfigsInNamespace(ctx context.Context, client k8sclient.Client,
...
+				PortSpecifier: &envoycore.SocketAddress_PortValue{
+					PortValue: uint32(containerPort),
+				},
+			},
+		},
+	}
+
+	cluster := envoyapi.Cluster{
+		Name:                 clusterName,
+		ConnectTimeout:       ptypes.DurationProto(2 * time.Second),
```
Comment (makslion) [inline pkg/resources/ratelimit/envoyconfig.go:194]: the only place time is used, but shouldn't it be a variable predefined along with others?

**Comment by makslion** [comment]
Verified code works and discussed code with Jose in the call. Comments are not related to the functionality of the code or to the simplification of the logic - only a convention and hard-wired "rules" in my head. Based on this:
/lgtm

File: pkg/resources/ratelimit/envoyconfig.go [hunk 15]
```diff
@@ -50,3 +71,187 @@ func DeleteEnvoyConfigsInNamespace(ctx context.Context, client k8sclient.Client,
...
+				socketAddress:
+					address: containerAddress
+					portValue: containerPort
+		name: apicast-ratelimit
+		type: STRICT_DNS
+**/
+func CreateClusterResource(containerAddress, clusterName string, containerPort int) *envoyapi.Cluster {
+
+	// Setting up cluster endpoints
+	clusterEndpoit := &envoycore.Address{
```
Comment (pmccarthy) [inline pkg/resources/ratelimit/envoyconfig.go:180]: Small typo on `clusterEndpoint `

File: pkg/resources/ratelimit/envoyconfig.go (hunk 14 above)
Comment (jjaferson) [reply pkg/resources/ratelimit/envoyconfig.go:194]: The timeout value should be defined depending on the time the application receiving the request will take to respond, for now, we are defaulting it to 2 seconds but if the timeout should be changed in the future, we can pass it as one of the parameters of the `CreateClusterResource` function

File: pkg/products/threescale/reconciler.go (hunk 8 above)
Comment (jjaferson) [reply pkg/products/threescale/reconciler.go:2158]: this route is created by the 3scale operator, so I am not sure if there is a constant with the service name there.

File: pkg/products/threescale/reconciler.go (hunk 9 above)
Comment (jjaferson) [reply pkg/products/threescale/reconciler.go:2193]: Good catch

**Pull request merged**